ruby/object_oriented_programming_basics/managing_ruby_projects.md

```diff
@@ -0,0 +1,239 @@
+
+puts "I'm blue da ba dee da ba di!".colorize(:blue)
+
+puts "It ain't easy bein' green...".colorize(:green)
+```
+
+You're probably itching to see all those colours, so run your file with `ruby foo.rb` to see them... or rather, a LoadError. Right - you need to <span id="install-gem">install that gem first!</span> Do that with `gem install colorize` and you'll see RubyGems in action. Your system now has access to the `Colorize` gem!
+
+Wait, *your system* - what about others that would like to use your code? Yeah, they would need to `gem install` it as well, no big deal. But what if you have dozens of gems? How do you make sure that the versions you use are the versions others download? This sounds rather tedious. Enter: Bundler. Another gem, part of RubyGems but released independently. <span id="bundler">Bundler allows you to declare what gems your project needs - down to their version.</span> As for others, Bundler allows them to take that declaration, a simple file called `Gemfile`, and use it to install those gems in a quick `bundle install`. Since gem installs are global, <span id="bundle-exec">you need a way to run only those particular gem versions that are declared in the `Gemfile`.</span> You can do that by using `bundle exec` followed by a command that you want to execute - most likely `bundle exec ruby foo.rb`.
+
+Let's make sure whoever wants to use our script can do so:
+
+```bash
+bundle init # creates a default Gemfile in the current working directory
+bundle add colorize # adds the colorize gem to the Gemfile and runs bundle install
+```
+
+This has created two files: `Gemfile` and `Gemfile.lock`. Let's take a look into both:
+
+```rb
+# Gemfile
+
+# frozen_string_literal: true
+
+source "https://rubygems.org"
+
+# gem "rails"
+
+gem "colorize", "~> 1.1"
+
+# Gemfile.lock
+GEM
+  remote: https://rubygems.org/
+  specs:
+    colorize (1.1.0)
+
+PLATFORMS
+  ruby
+  x86_64-linux # This might be different for you if you're using a different CPU and OS.
+
+DEPENDENCIES
+  colorize (~> 1.1)
+
+BUNDLED WITH
+   2.5.4
+```
+
+<span id="gemfile">There's not much in those but as you can see,</span> the `Gemfile` has information on where to get the gems from and what gems are required. The `"~> 1.1"` is something called a version constraint, particularly one called pessimistic constraint. It relies on semantic versioning - basically, the first number is the major version, second is the minor version and a third, if it exists is the patch number. Major versions can break things from previous versions - think changing method names for example. Minor versions can add and change things but they can't break anything. Patches happen when you introduce bug fixes that don't break anything. So, if people behind a gem maintain it in line with semantic versioning, you can rely on this pessimistic constraint never letting your project have a gem version that could potentially break your app - it is equivalent to `gem "colorize", ">= 1.1", "<2.0"`.
```

Comment:
To make the distinction between the different version numbers easier to visualise, we could have them broken out as bullet points:

```suggestion
<span id="gemfile">There's not much in those, but as you can see,</span> the `Gemfile` has information on where to get the gems from and what gems are required. 

The `"~> 1.1"` is a version constraint, particularly a pessimistic constraint. It relies on semantic versioning.

- The first number is the **major** version
- The second is the **minor** version
- the third, if it exists, is the **patch** number

Major versions can break things from previous versions - for example, changing method names. Minor versions can add and change things but can't break anything. Patches happen when you introduce bug fixes that don't break anything. 

So, if people behind a gem maintain it in line with semantic versioning, you can rely on this pessimistic constraint never letting your project have a gem version that could potentially break your app - it is equivalent to `gem "colorize", ">= 1.1", "<2.0"`.
```